table/block.cc

```diff
@@ -463,12 +487,77 @@ BlockIter* Block::NewIterator(const Comparator* cmp, const Comparator* ucmp,
     ret_iter->Invalidate(Status::OK());
     return ret_iter;
   } else {
+    const bool kKeyIncludesSeq = true;
+    ret_iter->InitializeBase(cmp, ucmp, data_, restart_offset_, num_restarts_,
+                             global_seqno_, kKeyIncludesSeq, cachable());
+  }
+
+  return ret_iter;
+}
+
+BlockIter* Block::NewIndexOrDataIterator(
+    const bool is_index, const Comparator* cmp, const Comparator* ucmp,
+    BlockIter* iter, Statistics* stats, bool total_order_seek,
+    bool key_includes_seq, BlockPrefixIndex* prefix_index) {
+  BlockIter* ret_iter;
+  if (iter != nullptr) {
+    ret_iter = iter;
+  } else {
+    if (is_index) {
+      ret_iter = new IndexBlockIter;
+    } else {
+      ret_iter = new DataBlockIter;
+    }
+  }
+  if (size_ < 2 * sizeof(uint32_t)) {
+    ret_iter->Invalidate(Status::Corruption("bad block contents"));
+    return ret_iter;
+  }
+  if (num_restarts_ == 0) {
+    // Empty block.
+    ret_iter->Invalidate(Status::OK());
+    return ret_iter;
+  } else if (is_index) {
     BlockPrefixIndex* prefix_index_ptr =
-        total_order_seek ? nullptr : prefix_index_.get();
-    ret_iter->Initialize(cmp, ucmp, data_, restart_offset_, num_restarts_,
-                         prefix_index_ptr, global_seqno_,
-                         read_amp_bitmap_.get(), key_includes_seq, cachable());
+        total_order_seek ? nullptr : prefix_index;
+    reinterpret_cast<IndexBlockIter*>(ret_iter)->Initialize(
+        cmp, ucmp, data_, restart_offset_, num_restarts_, prefix_index_ptr,
+        key_includes_seq, cachable());
```

Comment:
These casting is not ideal. I want to understand more about why we need them at all. Can we divide to two separate Block::NewIndexIterator() and Block::NewDataIterator()?

Reply:
This is because of partition indexes. Lots of code path is shared between index partition blocks and data blocks and they are distinguished with an is_index variable. We could avoid this NewIndexOrDataIterator function but it would not get rid of the casting; it just pushes it up one level to NewDataBlockIterator.